// @jbrowse/core/ui/DropDownMenu.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _Button = _interopRequireDefault(require("@material-ui/core/Button"));

var _styles = require("@material-ui/core/styles");

var _colorManipulator = require("@material-ui/core/styles/colorManipulator");

var _ArrowDropDown = _interopRequireDefault(require("@material-ui/icons/ArrowDropDown"));

var _mobxReact = require("mobx-react");

var _propTypes = _interopRequireDefault(require("prop-types"));

var _react = _interopRequireWildcard(require("react"));

var _Menu = _interopRequireDefault(require("./Menu"));

var useStyles = (0, _styles.makeStyles)(function (theme) {
  return {
    root: {
      display: 'flex'
    },
    buttonRoot: {
      '&:hover': {
        backgroundColor: (0, _colorManipulator.fade)(theme.palette.primary.contrastText, theme.palette.action.hoverOpacity),
        '@media (hover: none)': {
          backgroundColor: 'transparent'
        }
      }
    }
  };
});

function DropDownMenu(_ref) {
  var menuTitle = _ref.menuTitle,
      session = _ref.session,
      menuItems = _ref.menuItems;

  var _useState = (0, _react.useState)(false),
      _useState2 = (0, _slicedToArray2.default)(_useState, 2),
      open = _useState2[0],
      setOpen = _useState2[1];

  var anchorEl = (0, _react.useRef)(null);
  var classes = useStyles();

  function handleToggle() {
    setOpen(!open);
  }

  function handleMenuItemClick(_event, callback) {
    callback(session);
    handleClose();
  }

  function handleClose() {
    setOpen(false);
  }

  return /*#__PURE__*/_react.default.createElement("div", {
    className: classes.root
  }, /*#__PURE__*/_react.default.createElement(_Button.default, {
    ref: anchorEl,
    onClick: handleToggle,
    color: "inherit",
    "data-testid": "dropDownMenuButton",
    classes: {
      root: classes.buttonRoot
    }
  }, menuTitle, /*#__PURE__*/_react.default.createElement(_ArrowDropDown.default, null)), /*#__PURE__*/_react.default.createElement(_Menu.default, {
    anchorEl: anchorEl.current,
    anchorOrigin: {
      vertical: 'bottom',
      horizontal: 'left'
    },
    onMenuItemClick: handleMenuItemClick,
    open: open,
    onClose: handleClose,
    menuItems: menuItems
  }));
}

DropDownMenu.propTypes = {
  menuTitle: _propTypes.default.string.isRequired,
  menuItems: _mobxReact.PropTypes.arrayOrObservableArray.isRequired,
  session: _mobxReact.PropTypes.objectOrObservableObject.isRequired
};

var _default = (0, _mobxReact.observer)(DropDownMenu);

exports.default = _default;
