// @jbrowse/core/ui/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  LogoFull: true,
  Logomark: true,
  App: true,
  FileSelector: true,
  PrerenderedCanvas: true,
  ResizeHandle: true,
  EditableTypography: true,
  FactoryResetDialog: true,
  Tooltip: true,
  FatalErrorDialog: true,
  Menu: true,
  SanitizedHTML: true
};
Object.defineProperty(exports, "LogoFull", {
  enumerable: true,
  get: function get() {
    return _Logo.LogoFull;
  }
});
Object.defineProperty(exports, "Logomark", {
  enumerable: true,
  get: function get() {
    return _Logo.Logomark;
  }
});
Object.defineProperty(exports, "App", {
  enumerable: true,
  get: function get() {
    return _App.default;
  }
});
Object.defineProperty(exports, "FileSelector", {
  enumerable: true,
  get: function get() {
    return _FileSelector.default;
  }
});
Object.defineProperty(exports, "PrerenderedCanvas", {
  enumerable: true,
  get: function get() {
    return _PrerenderedCanvas.default;
  }
});
Object.defineProperty(exports, "ResizeHandle", {
  enumerable: true,
  get: function get() {
    return _ResizeHandle.default;
  }
});
Object.defineProperty(exports, "EditableTypography", {
  enumerable: true,
  get: function get() {
    return _EditableTypography.default;
  }
});
Object.defineProperty(exports, "FactoryResetDialog", {
  enumerable: true,
  get: function get() {
    return _FactoryResetDialog.default;
  }
});
Object.defineProperty(exports, "Tooltip", {
  enumerable: true,
  get: function get() {
    return _Tooltip.default;
  }
});
Object.defineProperty(exports, "FatalErrorDialog", {
  enumerable: true,
  get: function get() {
    return _FatalErrorDialog.default;
  }
});
Object.defineProperty(exports, "Menu", {
  enumerable: true,
  get: function get() {
    return _Menu.default;
  }
});
Object.defineProperty(exports, "SanitizedHTML", {
  enumerable: true,
  get: function get() {
    return _SanitizedHTML.default;
  }
});

var _theme = require("./theme");

Object.keys(_theme).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _theme[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _theme[key];
    }
  });
});

var _Logo = require("./Logo");

var _App = _interopRequireDefault(require("./App"));

var _FileSelector = _interopRequireDefault(require("./FileSelector"));

var _PrerenderedCanvas = _interopRequireDefault(require("./PrerenderedCanvas"));

var _ResizeHandle = _interopRequireDefault(require("./ResizeHandle"));

var _EditableTypography = _interopRequireDefault(require("./EditableTypography"));

var _FactoryResetDialog = _interopRequireDefault(require("./FactoryResetDialog"));

var _Tooltip = _interopRequireDefault(require("./Tooltip"));

var _FatalErrorDialog = _interopRequireDefault(require("./FatalErrorDialog"));

var _Menu = _interopRequireWildcard(require("./Menu"));

Object.keys(_Menu).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _Menu[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _Menu[key];
    }
  });
});

var _SanitizedHTML = _interopRequireDefault(require("./SanitizedHTML"));
