// @gmod/bam/dist/bamFile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = exports.BAM_MAGIC = void 0;

var _now = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/date/now"));

var _sort = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/sort"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/slicedToArray"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _promise = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/promise"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _reduce = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/reduce"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _slice = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/slice"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncToGenerator"));

var _bind = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/bind"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

var _wrapAsyncGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/wrapAsyncGenerator"));

var _awaitAsyncGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/awaitAsyncGenerator"));

var _asyncGeneratorDelegate2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncGeneratorDelegate"));

var _asyncIterator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncIterator"));

var _abortablePromiseCache = _interopRequireDefault(require("abortable-promise-cache"));

var _bai = _interopRequireDefault(require("./bai"));

var _csi = _interopRequireDefault(require("./csi"));

var _bufferCrc = _interopRequireDefault(require("buffer-crc32"));

var _bgzfFilehandle = require("@gmod/bgzf-filehandle");

var _object = _interopRequireDefault(require("object.entries-ponyfill"));

var _quickLru = _interopRequireDefault(require("quick-lru"));

var _genericFilehandle = require("generic-filehandle");

var _record = _interopRequireDefault(require("./record"));

var _sam = require("./sam");

var _util = require("./util");

var BAM_MAGIC = 21840194;
exports.BAM_MAGIC = BAM_MAGIC;
var blockLen = 1 << 16;

var BamFile = /*#__PURE__*/function () {
  /**
   * @param {object} args
   * @param {string} [args.bamPath]
   * @param {FileHandle} [args.bamFilehandle]
   * @param {string} [args.baiPath]
   * @param {FileHandle} [args.baiFilehandle]
   */
  function BamFile(_ref) {
    var _context;

    var bamFilehandle = _ref.bamFilehandle,
        bamPath = _ref.bamPath,
        bamUrl = _ref.bamUrl,
        baiPath = _ref.baiPath,
        baiFilehandle = _ref.baiFilehandle,
        baiUrl = _ref.baiUrl,
        csiPath = _ref.csiPath,
        csiFilehandle = _ref.csiFilehandle,
        csiUrl = _ref.csiUrl,
        cacheSize = _ref.cacheSize,
        fetchSizeLimit = _ref.fetchSizeLimit,
        chunkSizeLimit = _ref.chunkSizeLimit,
        _ref$yieldThreadTime = _ref.yieldThreadTime,
        yieldThreadTime = _ref$yieldThreadTime === void 0 ? 100 : _ref$yieldThreadTime,
        _ref$renameRefSeqs = _ref.renameRefSeqs,
        renameRefSeqs = _ref$renameRefSeqs === void 0 ? function (n) {
      return n;
    } : _ref$renameRefSeqs;
    (0, _classCallCheck2.default)(this, BamFile);
    (0, _defineProperty2.default)(this, "renameRefSeq", void 0);
    (0, _defineProperty2.default)(this, "bam", void 0);
    (0, _defineProperty2.default)(this, "index", void 0);
    (0, _defineProperty2.default)(this, "chunkSizeLimit", void 0);
    (0, _defineProperty2.default)(this, "fetchSizeLimit", void 0);
    (0, _defineProperty2.default)(this, "header", void 0);
    (0, _defineProperty2.default)(this, "featureCache", void 0);
    (0, _defineProperty2.default)(this, "chrToIndex", void 0);
    (0, _defineProperty2.default)(this, "indexToChr", void 0);
    (0, _defineProperty2.default)(this, "yieldThreadTime", void 0);
    this.renameRefSeq = renameRefSeqs;

    if (bamFilehandle) {
      this.bam = bamFilehandle;
    } else if (bamPath) {
      this.bam = new _genericFilehandle.LocalFile(bamPath);
    } else if (bamUrl) {
      this.bam = new _genericFilehandle.RemoteFile(bamUrl);
    } else {
      throw new Error('unable to initialize bam');
    }

    if (csiFilehandle) {
      this.index = new _csi.default({
        filehandle: csiFilehandle
      });
    } else if (csiPath) {
      this.index = new _csi.default({
        filehandle: new _genericFilehandle.LocalFile(csiPath)
      });
    } else if (csiUrl) {
      this.index = new _csi.default({
        filehandle: new _genericFilehandle.RemoteFile(csiUrl)
      });
    } else if (baiFilehandle) {
      this.index = new _bai.default({
        filehandle: baiFilehandle
      });
    } else if (baiPath) {
      this.index = new _bai.default({
        filehandle: new _genericFilehandle.LocalFile(baiPath)
      });
    } else if (baiUrl) {
      this.index = new _bai.default({
        filehandle: new _genericFilehandle.RemoteFile(baiUrl)
      });
    } else if (bamPath) {
      this.index = new _bai.default({
        filehandle: new _genericFilehandle.LocalFile("".concat(bamPath, ".bai"))
      });
    } else if (bamUrl) {
      this.index = new _bai.default({
        filehandle: new _genericFilehandle.RemoteFile("".concat(bamUrl, ".bai"))
      });
    } else {
      throw new Error('unable to infer index format');
    }

    this.featureCache = new _abortablePromiseCache.default({
      cache: new _quickLru.default({
        maxSize: cacheSize !== undefined ? cacheSize : 50
      }),
      fill: (0, _bind.default)(_context = this._readChunk).call(_context, this)
    });
    this.fetchSizeLimit = fetchSizeLimit || 500000000; // 500MB

    this.chunkSizeLimit = chunkSizeLimit || 300000000; // 300MB

    this.yieldThreadTime = yieldThreadTime;
  }

  (0, _createClass2.default)(BamFile, [{
    key: "getHeader",
    value: function () {
      var _getHeader = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var origOpts,
            opts,
            indexData,
            ret,
            buffer,
            res,
            bytesRead,
            uncba,
            headLen,
            _yield$this$_readRefS,
            chrToIndex,
            indexToChr,
            _args = arguments;

        return _regenerator.default.wrap(function _callee$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                origOpts = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};
                opts = (0, _util.makeOpts)(origOpts);
                _context2.next = 4;
                return this.index.parse(opts);

              case 4:
                indexData = _context2.sent;
                ret = indexData.firstDataLine ? indexData.firstDataLine.blockPosition + 65535 : undefined;

                if (!ret) {
                  _context2.next = 17;
                  break;
                }

                _context2.next = 9;
                return this.bam.read(Buffer.alloc(ret + blockLen), 0, ret + blockLen, 0, opts);

              case 9:
                res = _context2.sent;
                bytesRead = res.bytesRead;
                buffer = res.buffer;

                if (bytesRead) {
                  _context2.next = 14;
                  break;
                }

                throw new Error('Error reading header');

              case 14:
                if (bytesRead < ret) {
                  buffer = (0, _slice.default)(buffer).call(buffer, 0, bytesRead);
                } else {
                  buffer = (0, _slice.default)(buffer).call(buffer, 0, ret);
                }

                _context2.next = 20;
                break;

              case 17:
                _context2.next = 19;
                return this.bam.readFile(opts);

              case 19:
                buffer = _context2.sent;

              case 20:
                _context2.next = 22;
                return (0, _bgzfFilehandle.unzip)(buffer);

              case 22:
                uncba = _context2.sent;

                if (!(uncba.readInt32LE(0) !== BAM_MAGIC)) {
                  _context2.next = 25;
                  break;
                }

                throw new Error('Not a BAM file');

              case 25:
                headLen = uncba.readInt32LE(4);
                this.header = uncba.toString('utf8', 8, 8 + headLen);
                _context2.next = 29;
                return this._readRefSeqs(headLen + 8, 65535, opts);

              case 29:
                _yield$this$_readRefS = _context2.sent;
                chrToIndex = _yield$this$_readRefS.chrToIndex;
                indexToChr = _yield$this$_readRefS.indexToChr;
                this.chrToIndex = chrToIndex;
                this.indexToChr = indexToChr;
                return _context2.abrupt("return", (0, _sam.parseHeaderText)(this.header));

              case 35:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee, this);
      }));

      function getHeader() {
        return _getHeader.apply(this, arguments);
      }

      return getHeader;
    }()
  }, {
    key: "getHeaderText",
    value: function () {
      var _getHeaderText = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2() {
        var opts,
            _args2 = arguments;
        return _regenerator.default.wrap(function _callee2$(_context3) {
          while (1) {
            switch (_context3.prev = _context3.next) {
              case 0:
                opts = _args2.length > 0 && _args2[0] !== undefined ? _args2[0] : {};
                _context3.next = 3;
                return this.getHeader(opts);

              case 3:
                return _context3.abrupt("return", this.header);

              case 4:
              case "end":
                return _context3.stop();
            }
          }
        }, _callee2, this);
      }));

      function getHeaderText() {
        return _getHeaderText.apply(this, arguments);
      }

      return getHeaderText;
    }() // the full length of the refseq block is not given in advance so this grabs a chunk and
    // doubles it if all refseqs haven't been processed

  }, {
    key: "_readRefSeqs",
    value: function () {
      var _readRefSeqs2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee3(start, refSeqBytes) {
        var opts,
            res,
            bytesRead,
            buffer,
            uncba,
            nRef,
            p,
            chrToIndex,
            indexToChr,
            i,
            lName,
            refName,
            lRef,
            _args3 = arguments;
        return _regenerator.default.wrap(function _callee3$(_context4) {
          while (1) {
            switch (_context4.prev = _context4.next) {
              case 0:
                opts = _args3.length > 2 && _args3[2] !== undefined ? _args3[2] : {};

                if (!(start > refSeqBytes)) {
                  _context4.next = 3;
                  break;
                }

                return _context4.abrupt("return", this._readRefSeqs(start, refSeqBytes * 2, opts));

              case 3:
                _context4.next = 5;
                return this.bam.read(Buffer.alloc(refSeqBytes + blockLen), 0, refSeqBytes, 0, opts);

              case 5:
                res = _context4.sent;
                bytesRead = res.bytesRead;
                buffer = res.buffer;

                if (bytesRead) {
                  _context4.next = 10;
                  break;
                }

                throw new Error('Error reading refseqs from header');

              case 10:
                if (bytesRead < refSeqBytes) {
                  buffer = (0, _slice.default)(buffer).call(buffer, 0, bytesRead);
                } else {
                  buffer = (0, _slice.default)(buffer).call(buffer, 0, refSeqBytes);
                }

                _context4.next = 13;
                return (0, _bgzfFilehandle.unzip)(buffer);

              case 13:
                uncba = _context4.sent;
                nRef = uncba.readInt32LE(start);
                p = start + 4;
                chrToIndex = {};
                indexToChr = [];
                i = 0;

              case 19:
                if (!(i < nRef)) {
                  _context4.next = 35;
                  break;
                }

                _context4.next = 22;
                return (0, _util.abortBreakPoint)(opts.signal);

              case 22:
                lName = uncba.readInt32LE(p);
                refName = uncba.toString('utf8', p + 4, p + 4 + lName - 1);
                refName = this.renameRefSeq(refName);
                lRef = uncba.readInt32LE(p + lName + 4);
                chrToIndex[refName] = i;
                indexToChr.push({
                  refName: refName,
                  length: lRef
                });
                p = p + 8 + lName;

                if (!(p > uncba.length)) {
                  _context4.next = 32;
                  break;
                }

                console.warn("BAM header is very big.  Re-fetching ".concat(refSeqBytes, " bytes."));
                return _context4.abrupt("return", this._readRefSeqs(start, refSeqBytes * 2, opts));

              case 32:
                i += 1;
                _context4.next = 19;
                break;

              case 35:
                return _context4.abrupt("return", {
                  chrToIndex: chrToIndex,
                  indexToChr: indexToChr
                });

              case 36:
              case "end":
                return _context4.stop();
            }
          }
        }, _callee3, this);
      }));

      function _readRefSeqs(_x, _x2) {
        return _readRefSeqs2.apply(this, arguments);
      }

      return _readRefSeqs;
    }()
  }, {
    key: "getRecordsForRange",
    value: function () {
      var _getRecordsForRange = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee4(chr, min, max) {
        var opts,
            records,
            _iteratorNormalCompletion,
            _didIteratorError,
            _iteratorError,
            _iterator,
            _step,
            _value,
            chunk,
            _args4 = arguments;

        return _regenerator.default.wrap(function _callee4$(_context5) {
          while (1) {
            switch (_context5.prev = _context5.next) {
              case 0:
                opts = _args4.length > 3 && _args4[3] !== undefined ? _args4[3] : {
                  viewAsPairs: false,
                  pairAcrossChr: false,
                  maxInsertSize: 200000
                };
                records = [];
                _iteratorNormalCompletion = true;
                _didIteratorError = false;
                _context5.prev = 4;
                _iterator = (0, _asyncIterator2.default)(this.streamRecordsForRange(chr, min, max, opts));

              case 6:
                _context5.next = 8;
                return _iterator.next();

              case 8:
                _step = _context5.sent;
                _iteratorNormalCompletion = _step.done;
                _context5.next = 12;
                return _step.value;

              case 12:
                _value = _context5.sent;

                if (_iteratorNormalCompletion) {
                  _context5.next = 19;
                  break;
                }

                chunk = _value;
                records = (0, _concat.default)(records).call(records, chunk);

              case 16:
                _iteratorNormalCompletion = true;
                _context5.next = 6;
                break;

              case 19:
                _context5.next = 25;
                break;

              case 21:
                _context5.prev = 21;
                _context5.t0 = _context5["catch"](4);
                _didIteratorError = true;
                _iteratorError = _context5.t0;

              case 25:
                _context5.prev = 25;
                _context5.prev = 26;

                if (!(!_iteratorNormalCompletion && _iterator.return != null)) {
                  _context5.next = 30;
                  break;
                }

                _context5.next = 30;
                return _iterator.return();

              case 30:
                _context5.prev = 30;

                if (!_didIteratorError) {
                  _context5.next = 33;
                  break;
                }

                throw _iteratorError;

              case 33:
                return _context5.finish(30);

              case 34:
                return _context5.finish(25);

              case 35:
                return _context5.abrupt("return", records);

              case 36:
              case "end":
                return _context5.stop();
            }
          }
        }, _callee4, this, [[4, 21, 25, 35], [26,, 30, 34]]);
      }));

      function getRecordsForRange(_x3, _x4, _x5) {
        return _getRecordsForRange.apply(this, arguments);
      }

      return getRecordsForRange;
    }()
  }, {
    key: "streamRecordsForRange",
    value: function streamRecordsForRange(chr, min, max) {
      var _this = this;

      var opts = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : {
        viewAsPairs: false,
        pairAcrossChr: false,
        maxInsertSize: 200000
      };
      return (0, _wrapAsyncGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee5() {
        var _context7;

        var chrId, chunks, i, size, _context6, totalSize, _context8;

        return _regenerator.default.wrap(function _callee5$(_context9) {
          while (1) {
            switch (_context9.prev = _context9.next) {
              case 0:
                // todo regularize refseq names
                opts.viewAsPairs = opts.viewAsPairs || false;
                opts.pairAcrossChr = opts.pairAcrossChr || false;
                opts.maxInsertSize = opts.maxInsertSize !== undefined ? opts.maxInsertSize : 200000;
                chrId = _this.chrToIndex && _this.chrToIndex[chr];

                if (chrId >= 0) {
                  _context9.next = 8;
                  break;
                }

                chunks = [];
                _context9.next = 13;
                break;

              case 8:
                _context9.next = 10;
                return (0, _awaitAsyncGenerator2.default)(_this.index.blocksForRange(chrId, min - 1, max, opts));

              case 10:
                chunks = _context9.sent;

                if (chunks) {
                  _context9.next = 13;
                  break;
                }

                throw new Error('Error in index fetch');

              case 13:
                i = 0;

              case 14:
                if (!(i < chunks.length)) {
                  _context9.next = 23;
                  break;
                }

                _context9.next = 17;
                return (0, _awaitAsyncGenerator2.default)((0, _util.abortBreakPoint)(opts.signal));

              case 17:
                size = chunks[i].fetchedSize();

                if (!(size > _this.chunkSizeLimit)) {
                  _context9.next = 20;
                  break;
                }

                throw new Error((0, _concat.default)(_context6 = "Too many BAM features. BAM chunk size ".concat(size, " bytes exceeds chunkSizeLimit of ")).call(_context6, _this.chunkSizeLimit));

              case 20:
                i += 1;
                _context9.next = 14;
                break;

              case 23:
                totalSize = (0, _reduce.default)(_context7 = (0, _map.default)(chunks).call(chunks, function (s) {
                  return s.fetchedSize();
                })).call(_context7, function (a, b) {
                  return a + b;
                }, 0);

                if (!(totalSize > _this.fetchSizeLimit)) {
                  _context9.next = 26;
                  break;
                }

                throw new Error((0, _concat.default)(_context8 = "data size of ".concat(totalSize.toLocaleString(), " bytes exceeded fetch size limit of ")).call(_context8, _this.fetchSizeLimit.toLocaleString(), " bytes"));

              case 26:
                return _context9.delegateYield((0, _asyncGeneratorDelegate2.default)((0, _asyncIterator2.default)(_this._fetchChunkFeatures(chunks, chrId, min, max, opts)), _awaitAsyncGenerator2.default), "t0", 27);

              case 27:
              case "end":
                return _context9.stop();
            }
          }
        }, _callee5);
      }))();
    }
  }, {
    key: "_fetchChunkFeatures",
    value: function _fetchChunkFeatures(chunks, chrId, min, max, opts) {
      var _this2 = this;

      return (0, _wrapAsyncGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee6() {
        var featPromises, done, i, c, _yield$_awaitAsyncGen, data, cpositions, dpositions, chunk, promise, _i2;

        return _regenerator.default.wrap(function _callee6$(_context10) {
          while (1) {
            switch (_context10.prev = _context10.next) {
              case 0:
                featPromises = [];
                done = false;
                i = 0;

              case 3:
                if (!(i < chunks.length)) {
                  _context10.next = 21;
                  break;
                }

                c = chunks[i];
                _context10.next = 7;
                return (0, _awaitAsyncGenerator2.default)(_this2.featureCache.get(c.toString(), {
                  chunk: c,
                  opts: opts
                }, opts.signal));

              case 7:
                _yield$_awaitAsyncGen = _context10.sent;
                data = _yield$_awaitAsyncGen.data;
                cpositions = _yield$_awaitAsyncGen.cpositions;
                dpositions = _yield$_awaitAsyncGen.dpositions;
                chunk = _yield$_awaitAsyncGen.chunk;
                promise = _this2.readBamFeatures(data, cpositions, dpositions, chunk).then(function (records) {
                  var recs = [];

                  for (var _i = 0; _i < records.length; _i += 1) {
                    var feature = records[_i];

                    if (feature.seq_id() === chrId) {
                      if (feature.get('start') >= max) {
                        // past end of range, can stop iterating
                        done = true;
                        break;
                      } else if (feature.get('end') >= min) {
                        // must be in range
                        recs.push(feature);
                      }
                    }
                  }

                  return recs;
                });
                featPromises.push(promise);
                _context10.next = 16;
                return (0, _awaitAsyncGenerator2.default)(promise);

              case 16:
                if (!done) {
                  _context10.next = 18;
                  break;
                }

                return _context10.abrupt("break", 21);

              case 18:
                i++;
                _context10.next = 3;
                break;

              case 21:
                (0, _util.checkAbortSignal)(opts.signal);
                _i2 = 0;

              case 23:
                if (!(_i2 < featPromises.length)) {
                  _context10.next = 29;
                  break;
                }

                _context10.next = 26;
                return featPromises[_i2];

              case 26:
                _i2++;
                _context10.next = 23;
                break;

              case 29:
                (0, _util.checkAbortSignal)(opts.signal);

                if (!opts.viewAsPairs) {
                  _context10.next = 33;
                  break;
                }

                _context10.next = 33;
                return _this2.fetchPairs(chrId, featPromises, opts);

              case 33:
              case "end":
                return _context10.stop();
            }
          }
        }, _callee6);
      }))();
    }
  }, {
    key: "fetchPairs",
    value: function () {
      var _fetchPairs = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee10(chrId, featPromises, opts) {
        var _this3 = this,
            _context14,
            _context15;

        var unmatedPairs, readIds, matePromises, mateBlocks, mateChunks, i, mateTotalSize, _context16, mateFeatPromises, newMateFeats, featuresRet, newMates;

        return _regenerator.default.wrap(function _callee10$(_context18) {
          while (1) {
            switch (_context18.prev = _context18.next) {
              case 0:
                unmatedPairs = {};
                readIds = {};
                _context18.next = 4;
                return _promise.default.all((0, _map.default)(featPromises).call(featPromises, /*#__PURE__*/function () {
                  var _ref2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee7(f) {
                    var _context11;

                    var ret, readNames, i, name, id;
                    return _regenerator.default.wrap(function _callee7$(_context12) {
                      while (1) {
                        switch (_context12.prev = _context12.next) {
                          case 0:
                            _context12.next = 2;
                            return f;

                          case 2:
                            ret = _context12.sent;
                            readNames = {};

                            for (i = 0; i < ret.length; i++) {
                              name = ret[i].name();
                              id = ret[i].id();

                              if (!readNames[name]) {
                                readNames[name] = 0;
                              }

                              readNames[name]++;
                              readIds[id] = 1;
                            }

                            (0, _forEach.default)(_context11 = (0, _object.default)(readNames)).call(_context11, function (_ref3) {
                              var _ref4 = (0, _slicedToArray2.default)(_ref3, 2),
                                  k = _ref4[0],
                                  v = _ref4[1];

                              if (v === 1) {
                                unmatedPairs[k] = true;
                              }
                            });

                          case 6:
                          case "end":
                            return _context12.stop();
                        }
                      }
                    }, _callee7);
                  }));

                  return function (_x9) {
                    return _ref2.apply(this, arguments);
                  };
                }()));

              case 4:
                matePromises = [];
                _context18.next = 7;
                return _promise.default.all((0, _map.default)(featPromises).call(featPromises, /*#__PURE__*/function () {
                  var _ref5 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee8(f) {
                    var ret, i, name;
                    return _regenerator.default.wrap(function _callee8$(_context13) {
                      while (1) {
                        switch (_context13.prev = _context13.next) {
                          case 0:
                            _context13.next = 2;
                            return f;

                          case 2:
                            ret = _context13.sent;

                            for (i = 0; i < ret.length; i++) {
                              name = ret[i].name();

                              if (unmatedPairs[name] && (opts.pairAcrossChr || ret[i]._next_refid() === chrId && Math.abs(ret[i].get('start') - ret[i]._next_pos()) < (opts.maxInsertSize || 200000))) {
                                matePromises.push(_this3.index.blocksForRange(ret[i]._next_refid(), ret[i]._next_pos(), ret[i]._next_pos() + 1, opts));
                              }
                            }

                          case 4:
                          case "end":
                            return _context13.stop();
                        }
                      }
                    }, _callee8);
                  }));

                  return function (_x10) {
                    return _ref5.apply(this, arguments);
                  };
                }()));

              case 7:
                _context18.next = 9;
                return _promise.default.all(matePromises);

              case 9:
                mateBlocks = _context18.sent;
                mateChunks = [];

                for (i = 0; i < mateBlocks.length; i++) {
                  mateChunks = (0, _concat.default)(mateChunks).call(mateChunks, mateBlocks[i]);
                } // filter out duplicate chunks (the blocks are lists of chunks, blocks are concatenated, then filter dup chunks)


                mateChunks = (0, _filter.default)(_context14 = (0, _sort.default)(mateChunks).call(mateChunks)).call(_context14, function (item, pos, ary) {
                  return !pos || item.toString() !== ary[pos - 1].toString();
                });
                mateTotalSize = (0, _reduce.default)(_context15 = (0, _map.default)(mateChunks).call(mateChunks, function (s) {
                  return s.fetchedSize();
                })).call(_context15, function (a, b) {
                  return a + b;
                }, 0);

                if (!(mateTotalSize > this.fetchSizeLimit)) {
                  _context18.next = 16;
                  break;
                }

                throw new Error((0, _concat.default)(_context16 = "data size of ".concat(mateTotalSize.toLocaleString(), " bytes exceeded fetch size limit of ")).call(_context16, this.fetchSizeLimit.toLocaleString(), " bytes"));

              case 16:
                mateFeatPromises = (0, _map.default)(mateChunks).call(mateChunks, /*#__PURE__*/function () {
                  var _ref6 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee9(c) {
                    var _yield$_this3$feature, data, cpositions, dpositions, chunk, feats, mateRecs, _i3, feature;

                    return _regenerator.default.wrap(function _callee9$(_context17) {
                      while (1) {
                        switch (_context17.prev = _context17.next) {
                          case 0:
                            _context17.next = 2;
                            return _this3.featureCache.get(c.toString(), {
                              chunk: c,
                              opts: opts
                            }, opts.signal);

                          case 2:
                            _yield$_this3$feature = _context17.sent;
                            data = _yield$_this3$feature.data;
                            cpositions = _yield$_this3$feature.cpositions;
                            dpositions = _yield$_this3$feature.dpositions;
                            chunk = _yield$_this3$feature.chunk;
                            _context17.next = 9;
                            return _this3.readBamFeatures(data, cpositions, dpositions, chunk);

                          case 9:
                            feats = _context17.sent;
                            mateRecs = [];

                            for (_i3 = 0; _i3 < feats.length; _i3 += 1) {
                              feature = feats[_i3];

                              if (unmatedPairs[feature.get('name')] && !readIds[feature.id()]) {
                                mateRecs.push(feature);
                              }
                            }

                            return _context17.abrupt("return", mateRecs);

                          case 13:
                          case "end":
                            return _context17.stop();
                        }
                      }
                    }, _callee9);
                  }));

                  return function (_x11) {
                    return _ref6.apply(this, arguments);
                  };
                }());
                _context18.next = 19;
                return _promise.default.all(mateFeatPromises);

              case 19:
                newMateFeats = _context18.sent;
                featuresRet = [];

                if (newMateFeats.length) {
                  newMates = (0, _reduce.default)(newMateFeats).call(newMateFeats, function (result, current) {
                    return (0, _concat.default)(result).call(result, current);
                  });
                  featuresRet = (0, _concat.default)(featuresRet).call(featuresRet, newMates);
                }

                return _context18.abrupt("return", featuresRet);

              case 23:
              case "end":
                return _context18.stop();
            }
          }
        }, _callee10, this);
      }));

      function fetchPairs(_x6, _x7, _x8) {
        return _fetchPairs.apply(this, arguments);
      }

      return fetchPairs;
    }()
  }, {
    key: "_readChunk",
    value: function () {
      var _readChunk2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee11(_ref7, abortSignal) {
        var chunk, opts, c, bufsize, res, bytesRead, buffer, _yield$unzipChunkSlic, data, cpositions, dpositions;

        return _regenerator.default.wrap(function _callee11$(_context19) {
          while (1) {
            switch (_context19.prev = _context19.next) {
              case 0:
                chunk = _ref7.chunk, opts = _ref7.opts;
                c = chunk;
                bufsize = c.fetchedSize();
                _context19.next = 5;
                return this.bam.read(Buffer.alloc(bufsize), 0, bufsize, c.minv.blockPosition, opts);

              case 5:
                res = _context19.sent;
                bytesRead = res.bytesRead;
                buffer = res.buffer;
                (0, _util.checkAbortSignal)(abortSignal);

                if (bytesRead) {
                  _context19.next = 11;
                  break;
                }

                return _context19.abrupt("return", []);

              case 11:
                if (bytesRead < bufsize) {
                  buffer = (0, _slice.default)(buffer).call(buffer, 0, bytesRead);
                } else {
                  buffer = (0, _slice.default)(buffer).call(buffer, 0, bufsize);
                }

                _context19.next = 14;
                return (0, _bgzfFilehandle.unzipChunkSlice)(buffer, chunk);

              case 14:
                _yield$unzipChunkSlic = _context19.sent;
                data = _yield$unzipChunkSlic.buffer;
                cpositions = _yield$unzipChunkSlic.cpositions;
                dpositions = _yield$unzipChunkSlic.dpositions;
                (0, _util.checkAbortSignal)(abortSignal);
                return _context19.abrupt("return", {
                  data: data,
                  cpositions: cpositions,
                  dpositions: dpositions,
                  chunk: chunk
                });

              case 20:
              case "end":
                return _context19.stop();
            }
          }
        }, _callee11, this);
      }));

      function _readChunk(_x12, _x13) {
        return _readChunk2.apply(this, arguments);
      }

      return _readChunk;
    }()
  }, {
    key: "readBamFeatures",
    value: function () {
      var _readBamFeatures = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee12(ba, cpositions, dpositions, chunk) {
        var blockStart, sink, pos, last, blockSize, blockEnd, feature;
        return _regenerator.default.wrap(function _callee12$(_context20) {
          while (1) {
            switch (_context20.prev = _context20.next) {
              case 0:
                blockStart = 0;
                sink = [];
                pos = 0;
                last = +(0, _now.default)();

              case 4:
                if (!(blockStart + 4 < ba.length)) {
                  _context20.next = 18;
                  break;
                }

                blockSize = ba.readInt32LE(blockStart);
                blockEnd = blockStart + 4 + blockSize - 1; // increment position to the current decompressed status

                if (dpositions) {
                  while (blockStart + chunk.minv.dataPosition >= dpositions[pos++]) {}

                  pos--;
                } // only try to read the feature if we have all the bytes for it


                if (!(blockEnd < ba.length)) {
                  _context20.next = 15;
                  break;
                }

                feature = new _record.default({
                  bytes: {
                    byteArray: ba,
                    start: blockStart,
                    end: blockEnd
                  },
                  // the below results in an automatically calculated file-offset based ID
                  // if the info for that is available, otherwise crc32 of the features
                  //
                  // cpositions[pos] refers to actual file offset of a bgzip block boundaries
                  //
                  // we multiply by (1 <<8) in order to make sure each block has a "unique"
                  // address space so that data in that block could never overlap
                  //
                  // then the blockStart-dpositions is an uncompressed file offset from
                  // that bgzip block boundary, and since the cpositions are multiplied by
                  // (1 << 8) these uncompressed offsets get a unique space
                  //
                  // this has an extra chunk.minv.dataPosition added on because it blockStart
                  // starts at 0 instead of chunk.minv.dataPosition
                  //
                  // the +1 is just to avoid any possible uniqueId 0 but this does not realistically happen
                  fileOffset: cpositions ? cpositions[pos] * (1 << 8) + (blockStart - dpositions[pos]) + chunk.minv.dataPosition + 1 : _bufferCrc.default.signed((0, _slice.default)(ba).call(ba, blockStart, blockEnd))
                });
                sink.push(feature);

                if (!(this.yieldThreadTime && +(0, _now.default)() - last > this.yieldThreadTime)) {
                  _context20.next = 15;
                  break;
                }

                _context20.next = 14;
                return (0, _util.timeout)(1);

              case 14:
                last = +(0, _now.default)();

              case 15:
                blockStart = blockEnd + 1;
                _context20.next = 4;
                break;

              case 18:
                return _context20.abrupt("return", sink);

              case 19:
              case "end":
                return _context20.stop();
            }
          }
        }, _callee12, this);
      }));

      function readBamFeatures(_x14, _x15, _x16, _x17) {
        return _readBamFeatures.apply(this, arguments);
      }

      return readBamFeatures;
    }()
  }, {
    key: "hasRefSeq",
    value: function () {
      var _hasRefSeq = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee13(seqName) {
        var refId;
        return _regenerator.default.wrap(function _callee13$(_context21) {
          while (1) {
            switch (_context21.prev = _context21.next) {
              case 0:
                refId = this.chrToIndex && this.chrToIndex[seqName];
                return _context21.abrupt("return", this.index.hasRefSeq(refId));

              case 2:
              case "end":
                return _context21.stop();
            }
          }
        }, _callee13, this);
      }));

      function hasRefSeq(_x18) {
        return _hasRefSeq.apply(this, arguments);
      }

      return hasRefSeq;
    }()
  }, {
    key: "lineCount",
    value: function () {
      var _lineCount = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee14(seqName) {
        var refId;
        return _regenerator.default.wrap(function _callee14$(_context22) {
          while (1) {
            switch (_context22.prev = _context22.next) {
              case 0:
                refId = this.chrToIndex && this.chrToIndex[seqName];
                return _context22.abrupt("return", this.index.lineCount(refId));

              case 2:
              case "end":
                return _context22.stop();
            }
          }
        }, _callee14, this);
      }));

      function lineCount(_x19) {
        return _lineCount.apply(this, arguments);
      }

      return lineCount;
    }()
  }, {
    key: "indexCov",
    value: function () {
      var _indexCov = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee15(seqName, start, end) {
        var seqId;
        return _regenerator.default.wrap(function _callee15$(_context23) {
          while (1) {
            switch (_context23.prev = _context23.next) {
              case 0:
                _context23.next = 2;
                return this.index.parse();

              case 2:
                seqId = this.chrToIndex && this.chrToIndex[seqName];
                return _context23.abrupt("return", this.index.indexCov(seqId, start, end));

              case 4:
              case "end":
                return _context23.stop();
            }
          }
        }, _callee15, this);
      }));

      function indexCov(_x20, _x21, _x22) {
        return _indexCov.apply(this, arguments);
      }

      return indexCov;
    }()
  }]);
  return BamFile;
}();

exports.default = BamFile;
