// @gmod/bam/dist/chunk.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

// little class representing a chunk in the index
var Chunk = /*#__PURE__*/function () {
  /**
   * @param {VirtualOffset} minv
   * @param {VirtualOffset} maxv
   * @param {number} bin
   * @param {number} [fetchedSize]
   */
  function Chunk(minv, maxv, bin) {
    var fetchedSize = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : undefined;
    (0, _classCallCheck2.default)(this, Chunk);
    (0, _defineProperty2.default)(this, "minv", void 0);
    (0, _defineProperty2.default)(this, "maxv", void 0);
    (0, _defineProperty2.default)(this, "bin", void 0);
    (0, _defineProperty2.default)(this, "_fetchedSize", void 0);
    this.minv = minv;
    this.maxv = maxv;
    this.bin = bin;
    this._fetchedSize = fetchedSize;
  }

  (0, _createClass2.default)(Chunk, [{
    key: "toUniqueString",
    value: function toUniqueString() {
      var _context, _context2, _context3;

      return (0, _concat.default)(_context = (0, _concat.default)(_context2 = (0, _concat.default)(_context3 = "".concat(this.minv, "..")).call(_context3, this.maxv, " (bin ")).call(_context2, this.bin, ", fetchedSize ")).call(_context, this.fetchedSize(), ")");
    }
  }, {
    key: "toString",
    value: function toString() {
      return this.toUniqueString();
    }
  }, {
    key: "compareTo",
    value: function compareTo(b) {
      return this.minv.compareTo(b.minv) || this.maxv.compareTo(b.maxv) || this.bin - b.bin;
    }
  }, {
    key: "fetchedSize",
    value: function fetchedSize() {
      if (this._fetchedSize !== undefined) {
        return this._fetchedSize;
      }

      return this.maxv.blockPosition + (1 << 16) - this.minv.blockPosition;
    }
  }]);
  return Chunk;
}();

exports.default = Chunk;
