// @gmod/bam/dist/indexFile.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty2 = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty2(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/define-property"));

var _defineProperties = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/define-properties"));

var _getOwnPropertyDescriptors = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _getOwnPropertyDescriptor = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _getOwnPropertySymbols = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/keys"));

var _regenerator = _interopRequireDefault(require("@babel/runtime-corejs3/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/asyncToGenerator"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

var _defineProperty3 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

var _abortablePromiseCache = _interopRequireDefault(require("abortable-promise-cache"));

var _quickLru = _interopRequireDefault(require("quick-lru"));

function ownKeys(object, enumerableOnly) { var keys = (0, _keys.default)(object); if (_getOwnPropertySymbols.default) { var symbols = (0, _getOwnPropertySymbols.default)(object); if (enumerableOnly) symbols = (0, _filter.default)(symbols).call(symbols, function (sym) { return (0, _getOwnPropertyDescriptor.default)(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { var _context3; (0, _forEach.default)(_context3 = ownKeys(Object(source), true)).call(_context3, function (key) { (0, _defineProperty3.default)(target, key, source[key]); }); } else if (_getOwnPropertyDescriptors.default) { (0, _defineProperties.default)(target, (0, _getOwnPropertyDescriptors.default)(source)); } else { var _context4; (0, _forEach.default)(_context4 = ownKeys(Object(source))).call(_context4, function (key) { (0, _defineProperty2.default)(target, key, (0, _getOwnPropertyDescriptor.default)(source, key)); }); } } return target; }

var IndexFile = /*#__PURE__*/function () {
  /**
   * @param {filehandle} filehandle
   * @param {function} [renameRefSeqs]
   */
  function IndexFile(_ref) {
    var filehandle = _ref.filehandle,
        _ref$renameRefSeq = _ref.renameRefSeq,
        renameRefSeq = _ref$renameRefSeq === void 0 ? function (n) {
      return n;
    } : _ref$renameRefSeq;
    (0, _classCallCheck2.default)(this, IndexFile);
    (0, _defineProperty3.default)(this, "filehandle", void 0);
    (0, _defineProperty3.default)(this, "renameRefSeq", void 0);
    (0, _defineProperty3.default)(this, "_parseCache", void 0);
    this.filehandle = filehandle;
    this.renameRefSeq = renameRefSeq;
  }

  (0, _createClass2.default)(IndexFile, [{
    key: "_findFirstData",
    value: function _findFirstData(data, virtualOffset) {
      var currentFdl = data.firstDataLine;

      if (currentFdl) {
        data.firstDataLine = currentFdl.compareTo(virtualOffset) > 0 ? virtualOffset : currentFdl;
      } else {
        data.firstDataLine = virtualOffset;
      }
    }
  }, {
    key: "parse",
    value: function () {
      var _parse2 = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee() {
        var _this = this;

        var opts,
            _args = arguments;
        return _regenerator.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                opts = _args.length > 0 && _args[0] !== undefined ? _args[0] : {};

                if (!this._parseCache) {
                  this._parseCache = new _abortablePromiseCache.default({
                    cache: new _quickLru.default({
                      maxSize: 1
                    }),
                    fill: function fill(opts, signal) {
                      return _this._parse(_objectSpread(_objectSpread({}, opts), {}, {
                        signal: signal
                      }));
                    }
                  });
                }

                return _context.abrupt("return", this._parseCache.get('index', opts, opts.signal));

              case 3:
              case "end":
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function parse() {
        return _parse2.apply(this, arguments);
      }

      return parse;
    }()
  }, {
    key: "hasRefSeq",
    value: function () {
      var _hasRefSeq = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee2(seqId) {
        var opts,
            _args2 = arguments;
        return _regenerator.default.wrap(function _callee2$(_context2) {
          while (1) {
            switch (_context2.prev = _context2.next) {
              case 0:
                opts = _args2.length > 1 && _args2[1] !== undefined ? _args2[1] : {};
                _context2.next = 3;
                return this.parse(opts);

              case 3:
                _context2.t1 = seqId;
                _context2.t0 = _context2.sent.indices[_context2.t1];

                if (_context2.t0) {
                  _context2.next = 7;
                  break;
                }

                _context2.t0 = {};

              case 7:
                return _context2.abrupt("return", !!_context2.t0.binIndex);

              case 8:
              case "end":
                return _context2.stop();
            }
          }
        }, _callee2, this);
      }));

      function hasRefSeq(_x) {
        return _hasRefSeq.apply(this, arguments);
      }

      return hasRefSeq;
    }()
  }]);
  return IndexFile;
}();

exports.default = IndexFile;
