// @gmod/bam/dist/record.js
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault");

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");

_Object$defineProperty(exports, "__esModule", {
  value: true
});

exports.default = void 0;

var _flags2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/flags"));

var _parseInt2 = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/parse-int"));

var _filter = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/filter"));

var _keys = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/object/keys"));

var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));

var _concat = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/concat"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/createClass"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime-corejs3/helpers/defineProperty"));

var _map = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/map"));

var _constants = _interopRequireDefault(require("./constants"));

var _context;

var SEQRET_DECODER = (0, _map.default)(_context = '=ACMGRSVTWYHKDBN'.split('')).call(_context, function (s) {
  return s.charCodeAt(0);
});
var CIGAR_DECODER = 'MIDNSHP=X???????'.split('');
/**
 * Class of each BAM record returned by this API.
 */

var BamRecord = /*#__PURE__*/function () {
  function BamRecord(args) {
    (0, _classCallCheck2.default)(this, BamRecord);
    (0, _defineProperty2.default)(this, "data", void 0);
    (0, _defineProperty2.default)(this, "bytes", void 0);
    (0, _defineProperty2.default)(this, "flags", void 0);
    (0, _defineProperty2.default)(this, "_id", void 0);
    (0, _defineProperty2.default)(this, "_refID", void 0);
    (0, _defineProperty2.default)(this, "_tagOffset", undefined);
    (0, _defineProperty2.default)(this, "_tagList", []);
    (0, _defineProperty2.default)(this, "_allTagsParsed", false);
    this.data = {};
    this.bytes = {
      start: args.bytes.start,
      end: args.bytes.end,
      byteArray: args.bytes.byteArray
    };
    this._id = args.fileOffset;
    var _this$bytes = this.bytes,
        start = _this$bytes.start,
        byteArray = _this$bytes.byteArray;
    this._refID = byteArray.readInt32LE(start + 4);
    this.data.start = byteArray.readInt32LE(start + 8);
    this.flags = (byteArray.readInt32LE(start + 16) & 0xffff0000) >> 16;
  }

  (0, _createClass2.default)(BamRecord, [{
    key: "get",
    value: function get(field) {
      //@ts-ignore
      if (this[field]) {
        //@ts-ignore
        if (this.data[field]) {
          return this.data[field];
        } //@ts-ignore


        this.data[field] = this[field]();
        return this.data[field];
      }

      return this._get(field.toLowerCase());
    }
  }, {
    key: "end",
    value: function end() {
      return this.get('start') + (this.get('length_on_ref') || this.get('seq_length') || undefined);
    }
  }, {
    key: "seq_id",
    value: function seq_id() {
      return this._refID;
    } // same as get(), except requires lower-case arguments.  used
    // internally to save lots of calls to field.toLowerCase()

  }, {
    key: "_get",
    value: function _get(field) {
      if (field in this.data) {
        return this.data[field];
      }

      this.data[field] = this._parseTag(field);
      return this.data[field];
    }
  }, {
    key: "_tags",
    value: function _tags() {
      var _context2,
          _this = this;

      this._parseAllTags();

      var tags = ['seq'];

      if (!this.isSegmentUnmapped()) {
        tags.push('start', 'end', 'strand', 'score', 'qual', 'MQ', 'CIGAR', 'length_on_ref', 'template_length');
      }

      if (this.isPaired()) {
        tags.push('next_segment_position', 'pair_orientation');
      }

      tags = (0, _concat.default)(tags).call(tags, this._tagList || []);
      (0, _forEach.default)(_context2 = (0, _keys.default)(this.data)).call(_context2, function (k) {
        if (k[0] !== '_' && k !== 'next_seq_id') {
          tags.push(k);
        }
      });
      var seen = {};
      tags = (0, _filter.default)(tags).call(tags, function (t) {
        if (t in _this.data && _this.data[t] === undefined || t === 'CG' || t === 'cg') {
          return false;
        }

        var lt = t.toLowerCase();
        var s = seen[lt];
        seen[lt] = true;
        return !s;
      });
      return tags;
    }
  }, {
    key: "parent",
    value: function parent() {
      return undefined;
    }
  }, {
    key: "children",
    value: function children() {
      return this.get('subfeatures');
    }
  }, {
    key: "id",
    value: function id() {
      return this._id;
    } // special parsers

    /**
     * Mapping quality score.
     */

  }, {
    key: "mq",
    value: function mq() {
      var mq = (this.get('_bin_mq_nl') & 0xff00) >> 8;
      return mq === 255 ? undefined : mq;
    }
  }, {
    key: "score",
    value: function score() {
      return this.get('mq');
    }
  }, {
    key: "qual",
    value: function qual() {
      var _this$qualRaw;

      return (_this$qualRaw = this.qualRaw()) === null || _this$qualRaw === void 0 ? void 0 : _this$qualRaw.join(' ');
    }
  }, {
    key: "qualRaw",
    value: function qualRaw() {
      if (this.isSegmentUnmapped()) {
        return undefined;
      }

      var byteArray = this.bytes.byteArray;
      var p = this.bytes.start + 36 + this.get('_l_read_name') + this.get('_n_cigar_op') * 4 + this.get('_seq_bytes');
      var lseq = this.get('seq_length');
      var qseq = Buffer.allocUnsafe(lseq);

      for (var j = 0; j < lseq; ++j) {
        qseq[j] = byteArray[p + j];
      }

      return qseq;
    }
  }, {
    key: "strand",
    value: function strand() {
      return this.isReverseComplemented() ? -1 : 1;
    }
  }, {
    key: "multi_segment_next_segment_strand",
    value: function multi_segment_next_segment_strand() {
      if (this.isMateUnmapped()) {
        return undefined;
      }

      return this.isMateReverseComplemented() ? -1 : 1;
    }
  }, {
    key: "name",
    value: function name() {
      return this.get('_read_name');
    }
  }, {
    key: "_read_name",
    value: function _read_name() {
      var nl = this.get('_l_read_name');
      return this.bytes.byteArray.toString('ascii', this.bytes.start + 36, this.bytes.start + 36 + nl - 1);
    }
    /**
     * Get the value of a tag, parsing the tags as far as necessary.
     * Only called if we have not already parsed that field.
     */

  }, {
    key: "_parseTag",
    value: function _parseTag(tagName) {
      // if all of the tags have been parsed and we're still being
      // called, we already know that we have no such tag, because
      // it would already have been cached.
      if (this._allTagsParsed) {
        return undefined;
      }

      var byteArray = this.bytes.byteArray;
      var p = this._tagOffset || this.bytes.start + 36 + this.get('_l_read_name') + this.get('_n_cigar_op') * 4 + this.get('_seq_bytes') + this.get('seq_length');
      var blockEnd = this.bytes.end;
      var lcTag;

      while (p < blockEnd && lcTag !== tagName) {
        var tag = String.fromCharCode(byteArray[p], byteArray[p + 1]);
        lcTag = tag.toLowerCase();
        var type = String.fromCharCode(byteArray[p + 2]);
        p += 3;
        var value = void 0;

        switch (type) {
          case 'A':
            value = String.fromCharCode(byteArray[p]);
            p += 1;
            break;

          case 'i':
            value = byteArray.readInt32LE(p);
            p += 4;
            break;

          case 'I':
            value = byteArray.readUInt32LE(p);
            p += 4;
            break;

          case 'c':
            value = byteArray.readInt8(p);
            p += 1;
            break;

          case 'C':
            value = byteArray.readUInt8(p);
            p += 1;
            break;

          case 's':
            value = byteArray.readInt16LE(p);
            p += 2;
            break;

          case 'S':
            value = byteArray.readUInt16LE(p);
            p += 2;
            break;

          case 'f':
            value = byteArray.readFloatLE(p);
            p += 4;
            break;

          case 'Z':
          case 'H':
            value = '';

            while (p <= blockEnd) {
              var cc = byteArray[p++];

              if (cc === 0) {
                break;
              } else {
                value += String.fromCharCode(cc);
              }
            }

            break;

          case 'B':
            {
              value = '';
              var _cc = byteArray[p++];
              var Btype = String.fromCharCode(_cc);
              var limit = byteArray.readInt32LE(p);
              p += 4;

              if (Btype === 'i') {
                if (tag === 'CG') {
                  for (var k = 0; k < limit; k++) {
                    var cigop = byteArray.readInt32LE(p);
                    var lop = cigop >> 4;
                    var op = CIGAR_DECODER[cigop & 0xf];
                    value += lop + op;
                    p += 4;
                  }
                } else {
                  for (var _k = 0; _k < limit; _k++) {
                    value += byteArray.readInt32LE(p);

                    if (_k + 1 < limit) {
                      value += ',';
                    }

                    p += 4;
                  }
                }
              }

              if (Btype === 'I') {
                if (tag === 'CG') {
                  for (var _k2 = 0; _k2 < limit; _k2++) {
                    var _cigop = byteArray.readUInt32LE(p);

                    var _lop = _cigop >> 4;

                    var _op = CIGAR_DECODER[_cigop & 0xf];
                    value += _lop + _op;
                    p += 4;
                  }
                } else {
                  for (var _k3 = 0; _k3 < limit; _k3++) {
                    value += byteArray.readUInt32LE(p);

                    if (_k3 + 1 < limit) {
                      value += ',';
                    }

                    p += 4;
                  }
                }
              }

              if (Btype === 's') {
                for (var _k4 = 0; _k4 < limit; _k4++) {
                  value += byteArray.readInt16LE(p);

                  if (_k4 + 1 < limit) {
                    value += ',';
                  }

                  p += 2;
                }
              }

              if (Btype === 'S') {
                for (var _k5 = 0; _k5 < limit; _k5++) {
                  value += byteArray.readUInt16LE(p);

                  if (_k5 + 1 < limit) {
                    value += ',';
                  }

                  p += 2;
                }
              }

              if (Btype === 'c') {
                for (var _k6 = 0; _k6 < limit; _k6++) {
                  value += byteArray.readInt8(p);

                  if (_k6 + 1 < limit) {
                    value += ',';
                  }

                  p += 1;
                }
              }

              if (Btype === 'C') {
                for (var _k7 = 0; _k7 < limit; _k7++) {
                  value += byteArray.readUInt8(p);

                  if (_k7 + 1 < limit) {
                    value += ',';
                  }

                  p += 1;
                }
              }

              if (Btype === 'f') {
                for (var _k8 = 0; _k8 < limit; _k8++) {
                  value += byteArray.readFloatLE(p);

                  if (_k8 + 1 < limit) {
                    value += ',';
                  }

                  p += 4;
                }
              }

              break;
            }

          default:
            console.warn("Unknown BAM tag type '".concat(type, "', tags may be incomplete"));
            value = undefined;
            p = blockEnd;
          // stop parsing tags
        }

        this._tagOffset = p;

        this._tagList.push(tag);

        if (lcTag === tagName) {
          return value;
        }

        this.data[lcTag] = value;
      }

      this._allTagsParsed = true;
      return undefined;
    }
  }, {
    key: "_parseAllTags",
    value: function _parseAllTags() {
      this._parseTag('');
    }
  }, {
    key: "_parseCigar",
    value: function _parseCigar(cigar) {
      var _context3;

      return (//@ts-ignore
        (0, _map.default)(_context3 = cigar.match(/\d+\D/g)).call(_context3, function (op) {
          return [op.match(/\D/)[0].toUpperCase(), (0, _parseInt2.default)(op, 10)];
        })
      );
    }
    /**
     * @returns {boolean} true if the read is paired, regardless of whether both segments are mapped
     */

  }, {
    key: "isPaired",
    value: function isPaired() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FPAIRED);
    }
    /** @returns {boolean} true if the read is paired, and both segments are mapped */

  }, {
    key: "isProperlyPaired",
    value: function isProperlyPaired() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FPROPER_PAIR);
    }
    /** @returns {boolean} true if the read itself is unmapped; conflictive with isProperlyPaired */

  }, {
    key: "isSegmentUnmapped",
    value: function isSegmentUnmapped() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FUNMAP);
    }
    /** @returns {boolean} true if the read itself is unmapped; conflictive with isProperlyPaired */

  }, {
    key: "isMateUnmapped",
    value: function isMateUnmapped() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FMUNMAP);
    }
    /** @returns {boolean} true if the read is mapped to the reverse strand */

  }, {
    key: "isReverseComplemented",
    value: function isReverseComplemented() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FREVERSE);
    }
    /** @returns {boolean} true if the mate is mapped to the reverse strand */

  }, {
    key: "isMateReverseComplemented",
    value: function isMateReverseComplemented() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FMREVERSE);
    }
    /** @returns {boolean} true if this is read number 1 in a pair */

  }, {
    key: "isRead1",
    value: function isRead1() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FREAD1);
    }
    /** @returns {boolean} true if this is read number 2 in a pair */

  }, {
    key: "isRead2",
    value: function isRead2() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FREAD2);
    }
    /** @returns {boolean} true if this is a secondary alignment */

  }, {
    key: "isSecondary",
    value: function isSecondary() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FSECONDARY);
    }
    /** @returns {boolean} true if this read has failed QC checks */

  }, {
    key: "isFailedQc",
    value: function isFailedQc() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FQCFAIL);
    }
    /** @returns {boolean} true if the read is an optical or PCR duplicate */

  }, {
    key: "isDuplicate",
    value: function isDuplicate() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FDUP);
    }
    /** @returns {boolean} true if this is a supplementary alignment */

  }, {
    key: "isSupplementary",
    value: function isSupplementary() {
      return !!((0, _flags2.default)(this) & _constants.default.BAM_FSUPPLEMENTARY);
    }
  }, {
    key: "cigar",
    value: function cigar() {
      if (this.isSegmentUnmapped()) {
        return undefined;
      }

      var _this$bytes2 = this.bytes,
          byteArray = _this$bytes2.byteArray,
          start = _this$bytes2.start;
      var numCigarOps = this.get('_n_cigar_op');
      var p = start + 36 + this.get('_l_read_name');
      var seqLen = this.get('seq_length');
      var cigar = '';
      var lref = 0; // check for CG tag by inspecting whether the CIGAR field
      // contains a clip that consumes entire seqLen

      var cigop = byteArray.readInt32LE(p);
      var lop = cigop >> 4;
      var op = CIGAR_DECODER[cigop & 0xf];

      if (op === 'S' && lop === seqLen) {
        // if there is a CG the second CIGAR field will
        // be a N tag the represents the length on ref
        p += 4;
        cigop = byteArray.readInt32LE(p);
        lop = cigop >> 4;
        op = CIGAR_DECODER[cigop & 0xf];

        if (op !== 'N') {
          console.warn('CG tag with no N tag');
        }

        this.data.length_on_ref = lop;
        return this.get('CG');
      } else {
        for (var c = 0; c < numCigarOps; ++c) {
          cigop = byteArray.readInt32LE(p);
          lop = cigop >> 4;
          op = CIGAR_DECODER[cigop & 0xf];
          cigar += lop + op; // soft clip, hard clip, and insertion don't count toward
          // the length on the reference

          if (op !== 'H' && op !== 'S' && op !== 'I') {
            lref += lop;
          }

          p += 4;
        }

        this.data.length_on_ref = lref;
        return cigar;
      }
    }
  }, {
    key: "_flags",
    value: function _flags() {}
  }, {
    key: "length_on_ref",
    value: function length_on_ref() {
      this.get('cigar'); // the length_on_ref is set as a
      // side effect of the CIGAR parsing

      return this.data.length_on_ref;
    }
  }, {
    key: "_n_cigar_op",
    value: function _n_cigar_op() {
      return this.get('_flag_nc') & 0xffff;
    }
  }, {
    key: "_l_read_name",
    value: function _l_read_name() {
      return this.get('_bin_mq_nl') & 0xff;
    }
    /**
     * number of bytes in the sequence field
     */

  }, {
    key: "_seq_bytes",
    value: function _seq_bytes() {
      return this.get('seq_length') + 1 >> 1;
    }
  }, {
    key: "getReadBases",
    value: function getReadBases() {
      return this.seq();
    }
  }, {
    key: "seq",
    value: function seq() {
      var byteArray = this.bytes.byteArray;
      var p = this.bytes.start + 36 + this.get('_l_read_name') + this.get('_n_cigar_op') * 4;
      var seqBytes = this.get('_seq_bytes');
      var len = this.get('seq_length');
      var buf = '';
      var i = 0;

      for (var j = 0; j < seqBytes; ++j) {
        var sb = byteArray[p + j];
        buf += String.fromCharCode(SEQRET_DECODER[(sb & 0xf0) >> 4]);
        i++;

        if (i < len) {
          buf += String.fromCharCode(SEQRET_DECODER[sb & 0x0f]);
          i++;
        }
      }

      return buf;
    } // adapted from igv.js

  }, {
    key: "getPairOrientation",
    value: function getPairOrientation() {
      if (!this.isSegmentUnmapped() && !this.isMateUnmapped() && this._refID === this._next_refid()) {
        var s1 = this.isReverseComplemented() ? 'R' : 'F';
        var s2 = this.isMateReverseComplemented() ? 'R' : 'F';
        var o1 = ' ';
        var o2 = ' ';

        if (this.isRead1()) {
          o1 = '1';
          o2 = '2';
        } else if (this.isRead2()) {
          o1 = '2';
          o2 = '1';
        }

        var tmp = [];
        var isize = this.template_length();

        if (isize > 0) {
          tmp[0] = s1;
          tmp[1] = o1;
          tmp[2] = s2;
          tmp[3] = o2;
        } else {
          tmp[2] = s1;
          tmp[3] = o1;
          tmp[0] = s2;
          tmp[1] = o2;
        }

        return tmp.join('');
      }

      return null;
    }
  }, {
    key: "_bin_mq_nl",
    value: function _bin_mq_nl() {
      return this.bytes.byteArray.readInt32LE(this.bytes.start + 12);
    }
  }, {
    key: "_flag_nc",
    value: function _flag_nc() {
      return this.bytes.byteArray.readInt32LE(this.bytes.start + 16);
    }
  }, {
    key: "seq_length",
    value: function seq_length() {
      return this.bytes.byteArray.readInt32LE(this.bytes.start + 20);
    }
  }, {
    key: "_next_refid",
    value: function _next_refid() {
      return this.bytes.byteArray.readInt32LE(this.bytes.start + 24);
    }
  }, {
    key: "_next_pos",
    value: function _next_pos() {
      return this.bytes.byteArray.readInt32LE(this.bytes.start + 28);
    }
  }, {
    key: "template_length",
    value: function template_length() {
      return this.bytes.byteArray.readInt32LE(this.bytes.start + 32);
    }
  }, {
    key: "toJSON",
    value: function toJSON() {
      var _context4,
          _this2 = this;

      var data = {};
      (0, _forEach.default)(_context4 = (0, _keys.default)(this)).call(_context4, function (k) {
        if (k.charAt(0) === '_' || k === 'bytes') {
          return;
        } //@ts-ignore


        data[k] = _this2[k];
      });
      return data;
    }
  }]);
  return BamRecord;
}();

exports.default = BamRecord;
